### PR TITLE
[Bugfix][MRV2] Apply thinking budget to greedy requests

### PR DESCRIPTION
**Pull request by positive666**
## Summary

- Ensure Model Runner V2 does not take the no-logits-processing fast path when a scheduled request has an active thinking token budget.
- Add focused GPU regression coverage for greedy sampling.

## Problem

The Model Runner V2 sampler returns logits unchanged when `_requires_logits_processing()` is false. The predicate covered penalties, bad words, and non-default sampling, but did not cover `ThinkingBudgetState`.

As a result, a greedy request with only `thinking_token_budget` enabled could bypass `apply_thinking_budget()` entirely. This made the budget ineffective even though the request state was configured correctly.

This change includes active thinking-budget rows in the predicate. Requests without a thinking budget keep the existing fast path.

## Why this is not a duplicate

The thinking-budget implementation was added in #46727, while the logits-processing fast path was introduced in #47711. This PR fixes the interaction between those changes.

Immediately before submission, open-PR searches for `thinking_token_budget greedy` and `_requires_logits_processing` returned no matches. Broader sampler searches only returned speculative-decoding and frontend changes that do not modify this predicate.

## Tests

- `uv run ruff check vllm/v1/worker/gpu/sample/sampler.py tests/v1/worker/test_gpu_thinking_budget.py`
  - Passed
- `uv run ruff format --check vllm/v1/worker/gpu/sample/sampler.py tests/v1/worker/test_gpu_thinking_budget.py`
  - Passed
- `uv run pre-commit run mypy-3.12 --hook-stage manual --files vllm/v1/worker/gpu/sample/sampler.py tests/v1/worker/test_gpu_thinking_budget.py`
  - Passed
- `python -m pytest tests/v1/worker/test_gpu_thinking_budget.py -v` in the CUDA validation container
  - Baseline plus the new regression test: failed as expected because the end-reasoning logit remained `0.0` instead of `1.0e9`
  - Patched source: `13 passed`

## Model validation

Validated on a task-owned GPU with ModelScope Qwen3.5-9B:

- Greedy request without thinking returned the exact expected response.
- Greedy request with `thinking_token_budget=5` stopped reasoning and produced the final answer successfully.
- Existing service workloads on the other GPU were not modified.

The GPU source-overlay baseline was `ac7509e2`. The submitted base is `8e6d8e4f`; the intervening upstream change only touches XPU/release files and does not overlap this sampler or test. Static checks were rerun on the submitted base.

## AI assistance

AI assistance was used. The human submitter reviewed every changed line and the validation evidence and can defend the change end-to-end.

### CODE REVIEW

**Comment by Suppressor72** [comment]
Independent serve-level confirmation on dual RTX 5090, Qwen3.8-27B-FP8, TP=2, and MRV2.

Before this change, a greedy (`temperature=0`) request produced the exact same output with the thinking budget omitted and with `thinking_token_budget=16` (252 reasoning characters in each case). In a separate run with `speculative_config=None`, the unbounded, budget-16, and budget-256 requests were also identical (843 reasoning characters each), so speculative decoding was not required to reproduce the bypass.

I then manually applied only the predicate hunk proposed here to the installed production `sampler.py` and restarted the server. With C3 DSD/MTP still enabled, a paired greedy probe produced:

- Budget omitted: 491 locally counted reasoning tokens / 765 characters.
- `thinking_token_budget=16`: 15 reasoning tokens / 32 characters, cut mid-sentence.

Thus, the predicate hunk made the greedy thinking budget effective in this serve configuration.

This was not a checkout of the PR branch, and I did not run `tests/v1/worker/test_gpu_thinking_budget.py`; it is independent serve-level validation of the sampler change only.

**Comment by positive666** [comment]
Superseded by #52329, which carried this fix and has now been merged. Thanks @Suppressor72 for the independent validation!

**Comment by Suppressor72** [comment]
Follow-up to the confirmation above: I also observed the merged fix (`0.26.1rc1.dev926+gb05ae5dc0`, including #52329) enforce the budget under non-greedy sampling in one serve configuration.

Serve: Qwen3.6-35B-A3B-FP8, TP=2, MRV2, `--reasoning-parser qwen3`; sampling used `temperature=0.7, top_p=0.95` and streaming. The model's chat template contains no thinking-budget variable, so this was not template-injected.

- Budget omitted: the request reached `max_tokens=1024` while still in reasoning.
- `thinking_token_budget=64`: 63 locally counted reasoning tokens in both calls, followed by answer content.
- Across 12 seeded calls (four prompts × three seeds) at budgets 448, 4224, 4608, and 7296, locally retokenized streamed reasoning ended at exactly budget−1 in every call. Eleven finished `stop`; one budget-4224 call continued in content until the overall `max_tokens=16384` cap.

Scope: behavioral probes on one model and serve configuration.
